### PR TITLE
Kaixiang/pr/fsdp training improvements

### PR DESCRIPTION
**Pull request by nashjojo**
## Summary

Three FSDP2 training backend improvements: (1) fail-fast guard for undersized batches that cause NCCL deadlock, (2) internal micro-batch gradient accumulation to eliminate intra-step off-policy drift, and (3) order-preserving `_shard_list` that restores multi-actor support while guaranteeing deterministic sample ordering across ranks.

### 1. fix(fsdp): fail-fast when len(data) < world_size to avoid NCCL deadlock

When `fsdp_num_gpus >= 2` and input batch is smaller than world_size,
`_shard_list` produces empty shards. Empty-shard actors skip their NCCL
collectives, leaving other ranks blocked forever. Now raises `ValueError`
upfront so the future fails normally and the server stays healthy.

### 2. feat(fsdp): internal micro-batch grad accumulation

- `forward_backward` internally splits large batches into micro-batches,
  sharing one NCCL collective per RPC. Removes mini-batch SGD intra-step
  off-policy drift in PPO/GRPO-style RL training.
- `optimizer.zero_grad` deferred to `optim_step` to enable both internal
  micro-batch and cross-call gradient accumulation.
- `ModelConfig.fsdp_num_gpus`: documents single-GPU-only constraint inline.

### 3. feat(fsdp): restore multi-actor support with order-preserving _shard_list

- Replaces the round-robin shard assignment (which broke sample ordering
  across ranks) with a contiguous, order-preserving `_shard_list` that
  maintains deterministic sample order while distributing work evenly.
- Re-enables multi-actor FSDP (`fsdp_num_gpus >= 2`) that was temporarily
  disabled in commit 2 during the micro-batch refactor.
- Adds `test_fsdp_training_backend.py` with 105 lines of unit tests
  covering shard ordering correctness.

## Files Changed

- `src/tuft/backends/fsdp_training_backend.py` (+288/-47)
- `src/tuft/config.py` (+18/-6)
- `tests/test_fsdp_training_backend.py` (+105, new)

## Test Plan

- [x] `pytest tests/test_fsdp_training_backend.py` — shard ordering tests pass
- [x] Full pre-commit suite: ruff check, ruff format, pyright, detect-secrets (131 passed, 20 skipped)
- [x] Manual: multi-GPU FSDP2 training with `fsdp_num_gpus=2` on A100

### CODE REVIEW

File: src/tuft/config.py
Comment (linqinluli) [inline src/tuft/config.py:1]: Seems it's better to train with dp_size GPU when batch size > dp_size instead of force 1 dp_size.

File: src/tuft/backends/fsdp_training_backend.py [hunk 1]
```diff
@@ -122,6 +103,77 @@ def _chunk_tensordict_allow_2d_nested(td: TensorDict, chunks: int) -> list | tup
...
+    FSDPEngineWithLMHead as _FSDPEngineWithLMHead,
+)
+
+
+_v22_mb_state: Dict[str, Any] = {"count": 0, "start": 0.0}
+_v22_orig_forward_step = _FSDPEngineWithLMHead.forward_step  # type: ignore
+_v22_orig_forward_backward_batch = _FSDPEngineWithLMHead.forward_backward_batch
+
+
+def _v22_logged_forward_step(self, micro_batch, loss_function, forward_only):
```
Comment (linqinluli) [inline src/tuft/backends/fsdp_training_backend.py]: What's v22 related logging means? Seems it's a user specific debugging function.

File: src/tuft/backends/fsdp_training_backend.py [hunk 2]
```diff
@@ -215,10 +235,28 @@ def _datum_list_to_tensordict(
     td["input_ids"] = input_ids_nested
     td["position_ids"] = position_ids_nested
     tu.assign_non_tensor(td, global_token_num=global_token_num)
+    # v22 grad-accum: route through verl's `use_dynamic_bsz=False` path with
+    # a fixed `micro_batch_size_per_gpu`. verl's loop never zero_grad's, so
+    # gradients accumulate across all mb's; the caller's optim_step then
+    # consumes the accumulated grad and zero_grad's.
+    if micro_batch_size and micro_batch_size > 0 and batch_size % micro_batch_size == 0:
```
Comment (vanwaals) [inline src/tuft/backends/fsdp_training_backend.py:245]: when batch_size % micro_batch_size != 0, this quietly ignores the configured micro_batch_size and uses the full batch as a single micro-batch. This can silently cause OOM for large batches. Please add a logger.warning(...) when falling back, so operators can diagnose unexpected memory spikes.

File: tests/test_fsdp_training_backend.py [hunk 3]
```diff
@@ -310,3 +310,108 @@ async def _run_fsdp_single_process_flow() -> None:
...
+        assert output["datum_id"] == datum["id"], (
+            f"Order mismatch at index {i}: expected datum_id={datum['id']}, "
+            f"got {output['datum_id']}. This would cause silent logprob corruption "
+            f"in training (the [-am:] slicing hides shape mismatches)."
+        )
+        assert output["logprobs_len"] == len(datum["tokens"]), (
+            f"Length mismatch at index {i}: logprobs_len={output['logprobs_len']} "
+            f"!= tokens_len={len(datum['tokens'])}. This is the shape mismatch "
+            f"that Python's [-am:] slicing would hide."
+        )
```
Comment (vanwaals) [inline tests/test_fsdp_training_backend.py:417]: Missing coverage for the NCCL deadlock guard: there's no test verifying that forward() raises ValueError when len(data) < world_size in the multi-actor path. Consider adding a lightweight test (mock self._actors to a list of 2 stubs, call forward() with 1 datum, assert ValueError is raised with the expected message).

**Review by linqinluli — APPROVED** [review]
